core/lib/mini_merkle_tree/src/lib.rs

```diff
@@ -86,67 +118,135 @@ where
             hasher,
             hashes,
             binary_tree_size,
+            start_index: 0,
+            cache: vec![],
```

Comment:
Don't like this. As it's implemented right now, `cache` is the the Merkle path for the first retained element in the partial tree, unless the tree was never trimmed, or after it was entirely trimmed (AFAIU, in the second case, `cache` will contain garbage). These two corner case provide potential for bugs. IMO, it would make sense to:

- Either consistently compute `cache` so that it's *always* correctly defined. I.e., recompute it in the constructor. Regarding complete trimming, it's more tricky; you can compute the cache right away if you define it as a *subset* of a Merkle path to the left of the next tree element. In this case, I'd strongly suggest to signal this explicitly by defining a Merkle path subset as `Vec<Option<H256>>`. (For symmetry, the right path would be a subset of a Merkle path as well.) This would also make defining `cache` in the tree constructor trivial; it would be `vec![None; tree_depth]`.
- Or, make `cache` optional and recompute it on demand (probably would require some sort of interior mutability).

IMO, the first approach would be more appropriate here.